File: src/main/java/com/school/service/GradeService.java
package com.school.service;

import com.school.model.OptionalRequestParams;
import com.school.model.SubjectsWithGrades;
import com.schoolmodel.model.dto.StudentSubjectGradesDTO;
import com.school.repository.GradeRepository;
import com.school.repository.StudentRepository;
import com.school.repository.SubjectRepository;
import com.school.service.utils.mapper.QueryResultsMappingUtils;
import com.schoolmodel.model.entity.Grade;
import com.schoolmodel.model.entity.Student;
import com.schoolmodel.model.entity.Subject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@Service
public class GradeService {
    private static final Logger log = LoggerFactory.getLogger(GradeService.class);
    private final GradeRepository gradeRepository;
    private final StudentRepository studentRepository;
    private final SubjectRepository subjectRepository;

    public GradeService(GradeRepository gradeRepository, StudentRepository studentRepository, SubjectRepository subjectRepository) {
        this.gradeRepository = gradeRepository;
        this.studentRepository = studentRepository;
        this.subjectRepository = subjectRepository;
    }

    //TODO: refactor code to be more readable
    public Grade addGrade(int grade, Long subjectId, Long studentId) {
        log.info("Attempting to add grade {} to student's id {} to subjectId {}", grade, studentId, subjectId);
        Optional<Subject> subjectFound = subjectRepository.findById(subjectId);
        try {
            if (subjectFound.isPresent()) {
                Optional<Student> studentFound = studentRepository.findById(studentId);
                if (studentFound.isPresent()) {
                    log.info(studentFound.get().getName() + " " + studentFound.get().getSurname() + " found based passed id [" + studentFound + "]!");
                    if (grade > 0 && grade <= 6) {
                        Grade savedGrade = gradeRepository.save(new Grade(
                                grade,
                                studentFound.get(),
                                subjectFound.get())
                        );
                        log.debug("savedGrade");
                        log.debug(savedGrade.toString());
                        log.info("[Grade added!]");
                    } else {
                        String message = grade + " is improver value! Not going to be added";
                        log.error(message);
                        throw new IllegalArgumentException(message);
                    }
                } else {
                    String message = "Cannot find " + studentId + " student id to assign grade to!";
                    log.error(message);
                    throw new IllegalArgumentException(message);
                }
            } else {
                String message = "Cannot find " + subjectId + " subject to assign grade to!";
                log.error(message);
                throw new IllegalArgumentException(message);
            }
        } catch (Exception e) {
            log.error(e.getMessage());
            e.printStackTrace();
        }
        return null;
    }

    public List<StudentSubjectGradesDTO> getSubjectGradesForStudents(OptionalRequestParams params) {
        log.info("Getting grades for students grouped by subjects..");
        log.info("Params: {}", params);
        List<Object[]> results = gradeRepository.findAllGradesGroupedBySubject(
                params.getId(),
                params.getSubject(),
                params.getName(),
                params.getSurname(),
                params.getIdentifier()
        );

        try {
            List<StudentSubjectGradesDTO> grades = results.stream()
                    .map(QueryResultsMappingUtils::buildStudentSubjectGradesObject)
                    .toList();
            log.warn("Number of objects found : {}", grades.size());
            return populateWithExampleRecordIfListEmpty(grades);
        } catch (Exception e) {
            log.error(e.getMessage());
            log.error(Arrays.toString(e.getStackTrace()));
            throw e;
        }
    }

    public List<SubjectsWithGrades> getSubjectGradesForStudent(Long studentId) {
        log.debug("getSubjectGradesForStudent :: Student id param: {}", studentId);
        return gradeRepository.findAllGradesGroupedBySubjectForSingleStudent(studentId).stream()
                .map(QueryResultsMappingUtils::prepareGradesForSingleStudent)
                .toList();
    }

    private List<StudentSubjectGradesDTO> populateWithExampleRecordIfListEmpty(List<StudentSubjectGradesDTO> grades) {
        if (grades.isEmpty()) {
            return List.of(new StudentSubjectGradesDTO("No", "grades", "to", "display"));
        } else {
            return grades;
        }
    }
}

File: src/main/java/com/school/service/GradeService.kt
package com.school.service

import com.school.configuration.ApplicationConfig
import com.school.model.OptionalRequestParams
import com.school.model.SubjectsWithGrades
import com.school.model.dto.GradeDTO
import com.school.model.dto.StudentSubjectGradesDTO
import com.school.model.entity.Grade
import com.school.repository.sql.GradeRepository
import com.school.repository.sql.StudentRepository
import com.school.repository.sql.SubjectRepository
import com.school.service.utils.mapper.QueryResultsMapper.*
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class GradeService(
    private val subjectRepository: SubjectRepository,
    private val studentRepository: StudentRepository,
    private val sendNotificationToFrontendService: SendNotificationToFrontendService,
    private val gradeRepository: GradeRepository,
    private val applicationConfig: ApplicationConfig
) {

    private val log = LoggerFactory.getLogger(GradeService::class.java)

    fun addGrade(grade: GradeDTO): GradeDTO? {
        return try {
            log.info("GradeDTO to be saved: $grade")
            GradeDTO(gradeRepository.save(buildGradeWithCoupledEntityObjects(grade)))
                    .also { sendNotificationAboutGradeInsertStatus("OK") }
        } catch (e: Exception) {
            sendNotificationAboutGradeInsertStatus(e.message ?: "No grade added")
            log.error(e.message)
            throw RuntimeException(e)
        }
    }

    private fun buildGradeWithCoupledEntityObjects(grade: GradeDTO): Grade {
        return try {
            Grade().apply {
                gradeValue = validateGrade(grade.value)
                student = findStudentFromPassedGrade(grade.studentId)
                subject = findSubjectFromPassedGrade(grade.subjectId)
                note = grade.note
                gradeType = grade.gradeType
            }
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }
    }

    private fun validateGrade(gradeValue: Int) = gradeValue.takeIf { applicationConfig.availableGrades.contains(gradeValue) }
            ?: throw IllegalArgumentException("Grade value to add is not of available range")

    private fun findStudentFromPassedGrade(studentId: Long) = studentRepository.findById(studentId).takeIf { it.isPresent }?.get()
            ?: throw IllegalArgumentException("Cannot fund subject with id $studentId")

    private fun findSubjectFromPassedGrade(subjectId: Long) = subjectRepository.findById(subjectId).takeIf { it.isPresent }?.get()
            ?: throw IllegalArgumentException("Cannot fund subject with id $subjectId")

    private fun sendNotificationAboutGradeInsertStatus(message: String) {
        sendNotificationToFrontendService.notifyFrontendAboutGradeConsumptionStatus(message)
    }

    fun getSubjectGradesForStudents(params: OptionalRequestParams): List<StudentSubjectGradesDTO> {
        log.info("Params: {}", params)
        val queryResults = gradeRepository.findAllGradesGroupedBySubject(
                params.id,
                params.subject,
                params.name,
                params.surname,
                params.identifier
        )

        try {
            return queryResults
                    .map { result -> buildStudentSubjectGradesObject(result) }
                    .toList()
                    .takeIf { it.isNotEmpty() } ?: getEmptyGradesListIndicator()
        } catch (e: Exception) {
            log.error(e.message)
            log.error(e.stackTrace.contentToString())
            throw e
        }
    }

    fun getSubjectGradesForStudent(studentId: Long): List<SubjectsWithGrades> {
        return gradeRepository.findAllGradesGroupedBySubjectForSingleStudent(studentId)
                .map { result -> prepareGradesForSingleStudent(result) }
                .toList()
    }

    fun getGradeDetails(id: Long): GradeDTO {
        return gradeRepository.findById(id).also { log.info("Fetching grade details...") }
                .takeIf { it.isPresent }
                ?.get().also { log.info("Grade details found $it") }
                ?.let { GradeDTO(it) }
                ?: throw IllegalArgumentException("Could not find grade with ID $id")
    }
}